app/abci/handlers.go

```diff
@@ -0,0 +1,222 @@
+func (h *Handlers) PrepareProposalHandler() sdk.PrepareProposalHandler {
+	return func(ctx sdk.Context, req *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+		h.logger.Debug("start prepare proposal handler")
+
+		err := baseapp.ValidateVoteExtensions(ctx, h.stakingKeeper, req.Height, ctx.ChainID(), req.LocalLastCommit)
+		if err != nil {
+			return nil, err
+		}
+
+		proposalTxs := req.Txs
+		if req.Height > ctx.ConsensusParams().Abci.VoteExtensionsEnableHeight {
+			bz, err := json.Marshal(req.LocalLastCommit)
+			if err != nil {
+				h.logger.Error("failed to marshal extended votes", "err", err)
+			}
+			proposalTxs = append([][]byte{bz}, proposalTxs...)
+		}
+
+		return &abci.ResponsePrepareProposal{
+			Txs: proposalTxs,
+		}, nil
+	}
+}
+
+// ProcessProposalHandler handles the ProcessProposal ABCI to validate
+// the canonical set of vote extensions injected by the proposer.
+func (h *Handlers) ProcessProposalHandler() sdk.ProcessProposalHandler {
+	return func(ctx sdk.Context, req *abci.RequestProcessProposal) (*abci.ResponseProcessProposal, error) {
+		if req.Height > ctx.ConsensusParams().Abci.VoteExtensionsEnableHeight {
+			var extendedVotes abci.ExtendedCommitInfo
+			if err := json.Unmarshal(req.Txs[0], &extendedVotes); err != nil {
+				h.logger.Error("failed to decode injected extended votes tx", "err", err)
+				return &abci.ResponseProcessProposal{Status: abci.ResponseProcessProposal_REJECT}, nil
+			}
+
+			// Validate signatures and perform 2/3 voting power check.
+			err := baseapp.ValidateVoteExtensions(ctx, h.stakingKeeper, req.Height, ctx.ChainID(), extendedVotes)
+			if err != nil {
+				return nil, err
+			}
+
+			// Verify every batch signature.
+			batch, err := h.batchingKeeper.GetBatchForHeight(ctx, ctx.BlockHeight()-2)
+			if err != nil {
+				return nil, err
+			}
+			for _, vote := range extendedVotes.Votes {
+				ok, err := h.verifyBatchSignature(ctx, batch.BatchId, vote.VoteExtension[:64], vote.Validator.Address, utils.Secp256k1)
+				if !ok || err != nil {
+					return nil, err
+				}
+			}
+		}
+
+		return &abci.ResponseProcessProposal{Status: abci.ResponseProcessProposal_ACCEPT}, nil
+	}
+}
+
+// PreBlocker runs before BeginBlocker to extract the batch signatures
+// from the canonical set of vote extensions injected by the proposer
+// and store them.
+func (h *Handlers) PreBlocker() sdk.PreBlocker {
+	return func(ctx sdk.Context, req *abci.RequestFinalizeBlock) (res *sdk.ResponsePreBlock, err error) {
+		// Use defer to prevent returning an error, which would cause
+		// the chain to halt.
+		defer func() {
+			// Handle a panic.
+			if r := recover(); r != nil {
+				h.logger.Error("recovered from panic in pre-blocker", "err", r)
+			}
+			// Handle an error.
+			if err != nil {
+				h.logger.Error("error in pre-blocker", "err", err)
+			}
+			err = nil
+		}()
+
+		res = new(sdk.ResponsePreBlock)
+		if req.Height > ctx.ConsensusParams().Abci.VoteExtensionsEnableHeight {
```

Comment:
Maybe an early return makes sense here?

Reply:
Where?

Reply:
Instead of having all the relevant code in the `if` clause we can do 

```golang
if req.Height <= ctx.ConsensusParams().Abci.VoteExtensionsEnableHeight {
  return res, nil
}
```